fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -216,6 +232,251 @@ public List<OptExpression> rewrite(ReplaceColumnRefRewriter queryColumnRefRewrit
+            return false;
+        }
+        if (mvTableScanDescs.stream().anyMatch(tableScanDesc -> !tableScanDesc.getTable().isNativeTable())) {
+            return false;
+        }
+
+        Multimap<String, TableScanDesc> nameToTable = ArrayListMultimap.create();
+        for (TableScanDesc tableScanDesc : mvTableScanDescs) {
+            graph.addNode(tableScanDesc);
+            nameToTable.put(tableScanDesc.getName(), tableScanDesc);
+            if (!queryTableScanDescs.contains(tableScanDesc)) {
+                extraTableScanDescs.add(tableScanDesc);
+            }
+        }
+        Map<TableScanDesc, List<ColumnRefOperator>> extraTableColumns = Maps.newHashMap();
+
+        // add edges to directed graph by FK-UK
+        for (TableScanDesc tableScanDesc : graph.nodes()) {
+            // now only support OlapTable
+            Preconditions.checkState(tableScanDesc.getTable() instanceof OlapTable);
+            OlapTable childTable = (OlapTable) tableScanDesc.getTable();
+            List<ForeignKeyConstraint> foreignKeyConstraints = childTable.getForeignKeyConstraints();
+            if (foreignKeyConstraints == null) {
+                continue;
+            }
+
+            for (ForeignKeyConstraint foreignKeyConstraint : foreignKeyConstraints) {
+                Collection<TableScanDesc> parentTableScanDescs =
+                        nameToTable.get(foreignKeyConstraint.getParentTableInfo().getTableName());
+                if (parentTableScanDescs == null || parentTableScanDescs.isEmpty()) {
+                    continue;
+                }
+                List<Pair<String, String>> columnPairs = foreignKeyConstraint.getColumnRefPairs();
+                List<String> childKeys = columnPairs.stream().map(pair -> pair.first).collect(Collectors.toList());
+                List<String> parentKeys = columnPairs.stream().map(pair -> pair.second).collect(Collectors.toList());
+                for (TableScanDesc parentTableScanDesc : parentTableScanDescs) {
+                    List<ColumnRefOperator> tableCompensationColumns = Lists.newArrayList();
+                    Multimap<ColumnRefOperator, ColumnRefOperator> tmpCompensationJoinColumns = ArrayListMultimap.create();
+                    if (!extraJoinCheck(parentTableScanDesc, tableScanDesc, columnPairs, childKeys, parentKeys,
+                            viewEquivalenceClasses, tableCompensationColumns, tmpCompensationJoinColumns)) {
+                        continue;
+                    }
+
+                    if (extraTableScanDescs.contains(parentTableScanDesc)) {
+                        extraTableColumns.put(parentTableScanDesc, tableCompensationColumns);
+                        compensationJoinColumns.putAll(tmpCompensationJoinColumns);
+                    }
+                    // add an edge to graph
+                    if (!graph.successors(tableScanDesc).contains(parentTableScanDesc)) {
```

Comment:
if `graph.successors(tableScanDesc).contains(parentTableScanDesc)` is true, just break? no need loop?

Reply:
no, we should check every parentTableScanDescs.
for a join b1 join b2, we should try to add both b1 and b2 into the graph